### DISCUSSION TITLE
cy.exec in command.ts file

### DISCUSSION THREAD

**Discussion by gDev95** [cypress-io/cypress#28096] [Questions and Help]
I upgraded Cypress from v10.x.x to 12.x and then 13.x but my tests fail because I am running `cy.exec` outside a `spec` file. 

Why the hell would that matter. What is the supposed workaround for such use cases? 

e.g 

```
Cypress.Commands.add('moveTicketToXXXState', function (ticketId) {
  cy.exec(`npm run run-e2e-setup ${ticketId}`);
  cy.readFile('response.json').then((result) => {
    if (typeof result !== 'boolean') {
      throw new Error('Error occured');
    }
  });
});
```

**Comment by portyu9** [comment]
Hi @gDev95 

For Cypress 12/13, the fact that the custom command is **defined in `commands.ts` is not itself a problem**. Custom commands are normally defined there.

The important part is when the command executes: Cypress commands can only execute while a test or hook is running. So if this custom command (or a helper that calls `cy.exec()`) is invoked at module/top level, Cypress will report that there is no active test.

For current Cypress there is also an important change: **`cy.exec()` was removed in Cypress 16**.

The supported replacement is to run the external operation in Node through `cy.task()`.

For example:

```js
// cypress.config.js
const { defineConfig } = require('cypress')
const { exec } = require('node:child_process')
const { promisify } = require('node:util')
const fs = require('node:fs/promises')

const execAsync = promisify(exec)

module.exports = defineConfig({
  e2e: {
    setupNodeEvents(on, config) {
      on('task', {
        async moveTicket(ticketId) {
          await execAsync(`npm run run-e2e-setup -- ${ticketId}`)

          const result = JSON.parse(
            await fs.readFile('response.json', 'utf8')
          )

          if (typeof result !== 'boolean') {
            throw new Error('Error occurred')
          }

          return result
        },
      })
    },
  },
})
```

Then your custom command can simply call the task:

```js
Cypress.Commands.add('moveTicketToXXXState', (ticketId) => {
  return cy.task('moveTicket', ticketId)
})
```

If `run-e2e-setup` is JavaScript you control, it would be even cleaner to import/call that Node function directly from the task instead of launching another npm process.

Please mark this answer as accepted if it helped, thank you.
